src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,361 @@
+# TODO:
+# * add transform and transfom! functions
+# * add `Col` wrapper for whole column operations
+# * update documentation
+# * add tests
+
+# normalize_selection function makes sure that whatever input format of idx is it
+# will end up in one of four canonical forms
+# 1) Int
+# 2) AbstractVector{Int}
+# 3) Pair{Int, Pair{ColRename, Symbol}}
+# 3) Pair{Int, <:Pair{<:Base.Callable, Symbol}}
+# 4) Pair{AbstractVector{Int}, <:Pair{<:Base.Callable, Symbol}}
+# in this way we can easily later decide on the codepath using type signatures
+
+```
+    ColRename
+
+A singleton type indicating that column renaming operation was requested in `select`.
+```
+struct ColRename end
+
+normalize_selection(idx::AbstractIndex, sel) = idx[sel]
+normalize_selection(idx::AbstractIndex, sel::Pair{<:Any,<:Pair{<:Base.Callable,Symbol}}) =
+    idx[first(sel)] => last(sel)
+normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, <:Symbol}) =
+    idx[first(sel)] => ColRename() => last(sel)
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:ColumnIndex, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    newcol = Symbol(_names(idx)[c], "_", funname(fun))
+    return c => fun => newcol
+end
+
+function normalize_selection(idx::AbstractIndex, sel::Pair{<:Any, <:Base.Callable})
+    c = idx[first(sel)]
+    fun = last(sel)
+    if length(c) > 2
+        newcol = Symbol(join(_names(idx)[c[1:2]], "_"), "_⋯_", funname(fun))
+    else
+        newcol = Symbol(join(_names(idx)[c], "_"), "_", funname(fun))
+    end
+    return c => fun => newcol
+end
+
+struct TypeHolder{T} end
+
+function select_transform_helper(th::TypeHolder{T}, cols, fun, n) where T
```

Comment:
@nalimilan + @quinnj Do you maybe know a cleaner way to generate a type-stable iterator that produces `NamedTuple`s from a collection of columns. In short I want `map` in line 51 below to be able to correctly infer the return type of `fun` when it gets rows of some data frame as `NamedTuples`, but I want to avoid materializing `Tables.rowtable` as it is expensive.

What I have implemented is inferred by the compiler but I feel that the design here is not clean and maybe something simpler is possible.

Reply:
In Tables.jl, we have the `NamedTupleIterator`, which you could use like `r = Tables.rows(df); Tables.namedtupleiterator(eltype(r), r)`.

Reply:
I'd have simply called `map(fun, Tables.rows(Tables.columntable(df[cols])))` (or a more efficient version which directly builds the named tuple of vectors). Isn't that enough?

Reply:
Well, `Tables.rows` on `columntable` doesn't produce NamedTuples. It does a `ColumnsRow`. It might still work though and would indeed be a bit simpler.

Reply:
`Tables.namedtupleiterator` is what I needed (now I see that it essentially does what I do but in a cleaner and more generic way - thank you). Is this function documented (i.e. a public API of Tables.jl). I will use it anyway of course.

The whole pain of the request is that we want to pass `NamedTuple` exactly (this is what we have agreed on with @nalimilan).

The correct code that should be used (and this is what I will add in the implementation) is:
```
r = Tables.rows(Tables.columntable(df))
map(fun, Tables.namedtupleiterator(eltype(r), r))
```
(only this approach ensures that the code is fast)

Thank you for the support.

Reply:
`Tables.namedtupleiterator` isn't publicly documented, but I think it's fine to use. I think that all sounds good to me.

Reply:
Indeed.

I imagine the name `ColumnsRow` is going to scare quite a few users... :-p